File: dp_labs_and_solutions/dp_labs_and_solutions/dp_java_labs_and_solutions/decorator/solution/With.java
// With.java

public abstract class With extends Beverage
{
	public With(Beverage aSuccessor)
	{
		setSuccessor(aSuccessor);
	}

	public void prepare()
	{
		getSuccessor().prepare();
	}

	private Beverage getSuccessor()
	{
		return mySuccessor;
	}

	private void setSuccessor(Beverage aSuccessor)
	{
		if (aSuccessor == null)
		{
			// throw an exception.
		}
		mySuccessor = aSuccessor;
	}

	private Beverage mySuccessor;
}

File: dp_labs_and_solutions/dp_labs_and_solutions/dp_cpp_labs_and_solutions/decorator/solution/with.cpp
// with.cpp

#include <cstdlib>
class beverage;
#include "with.h"

with::with(beverage * a_successor)
{
	set_successor(a_successor);
}

void with::prepare()
{
    get_successor()->prepare();
}

beverage * with::get_successor() const
{
	return my_successor;
}

void with::set_successor(beverage * a_successor)
{
	if (a_successor == nullptr)
	{
		// throw an exception.
	}
	my_successor = a_successor;
}